File: go/forth/forth.go
package forth

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
)

// Define errors
var (
	ErrDivisionByZero = errors.New("division by zero")
	ErrStackUnderflow = errors.New("stack underflow")
	ErrUnknownWord    = errors.New("unknown word")
	ErrInvalidWord    = errors.New("invalid word definition")
)

// forthState holds the stack and custom word definitions.
type forthState struct {
	stack       []int
	definitions map[string][]string
	defining    bool
	currentDef  []string
	currentName string
}

// newForthState creates a new Forth state.
func newForthState() *forthState {
	// Initialize built-in definitions (although handled directly in eval for performance)
	defs := make(map[string][]string)
	return &forthState{
		stack:       make([]int, 0),
		definitions: defs,
		defining:    false,
	}
}

// push adds an element to the stack.
func (s *forthState) push(val int) {
	s.stack = append(s.stack, val)
}

// pop removes and returns the top element from the stack.
func (s *forthState) pop() (int, error) {
	if len(s.stack) == 0 {
		return 0, ErrStackUnderflow
	}
	val := s.stack[len(s.stack)-1]
	s.stack = s.stack[:len(s.stack)-1]
	return val, nil
}

// eval evaluates a sequence of words iteratively.
func (s *forthState) eval(initialWords []string) error {
	// Use a slice as a stack for words to process. Start with the initial words.
	// Create a copy to avoid modifying the original slice from the caller (e.g., definitions map)
	wordsToProcess := make([]string, len(initialWords))
	copy(wordsToProcess, initialWords)

	for len(wordsToProcess) > 0 {
		// Pop the next word from the front
		word := wordsToProcess[0]
		wordsToProcess = wordsToProcess[1:] // Consume the word

		lowerWord := strings.ToLower(word)

		if s.defining {
			if lowerWord == ";" {
				if s.currentName == "" { // Name must be set after ':'
					s.defining = false // Reset state
					s.currentName = ""
					s.currentDef = nil
					return ErrInvalidWord
				}
				// Check for redefinition of numbers - not allowed
				if _, err := strconv.Atoi(s.currentName); err == nil {
					s.defining = false // Reset state
					s.currentName = ""
					s.currentDef = nil
					return ErrInvalidWord
				}
				// Store the definition
				s.definitions[s.currentName] = s.currentDef
				s.defining = false
				s.currentName = ""
				s.currentDef = nil
			} else if s.currentName == "" { // This is the word name after ':'
				// Check if the word name is a number - not allowed
				if _, err := strconv.Atoi(lowerWord); err == nil {
					s.defining = false // Reset state
					return ErrInvalidWord
				}
				s.currentName = lowerWord // Store definition name in lower case
			} else { // This is part of the definition body
				s.currentDef = append(s.currentDef, word) // Store original case word in definition
			}
			continue // Move to the next word in the input sequence
		}

		// Try parsing as a number
		if val, err := strconv.Atoi(word); err == nil {
			s.push(val)
			continue
		}

		// Handle built-in words and custom definitions
		switch lowerWord {
		case ":":
			if s.defining {
				return ErrInvalidWord // Cannot nest definitions
			}
			s.defining = true
			s.currentName = "" // Reset, will be set by the next word
			s.currentDef = make([]string, 0)
		case "+":
			if len(s.stack) < 2 { return ErrStackUnderflow }
			b, _ := s.pop()
			a, _ := s.pop()
			s.push(a + b)
		case "-":
			if len(s.stack) < 2 { return ErrStackUnderflow }
			b, _ := s.pop()
			a, _ := s.pop()
			s.push(a - b)
		case "*":
			if len(s.stack) < 2 { return ErrStackUnderflow }
			b, _ := s.pop()
			a, _ := s.pop()
			s.push(a * b)
		case "/":
			if len(s.stack) < 2 { return ErrStackUnderflow }
			b, _ := s.pop()
			a, _ := s.pop()
			if b == 0 { return ErrDivisionByZero }
			s.push(a / b)
		case "dup":
			if len(s.stack) < 1 { return ErrStackUnderflow }
			val := s.stack[len(s.stack)-1]
			s.push(val)
		case "drop":
			if _, err := s.pop(); err != nil { return err }
		case "swap":
			if len(s.stack) < 2 { return ErrStackUnderflow }
			b, _ := s.pop()
			a, _ := s.pop()
			s.push(b)
			s.push(a)
		case "over":
			if len(s.stack) < 2 { return ErrStackUnderflow }
			a := s.stack[len(s.stack)-2]
			s.push(a)
		default:
			// Check if it's a custom defined word
			if def, ok := s.definitions[lowerWord]; ok {
				// Prepend the definition's words to the list of words to process.
				// Make a copy of the definition slice to avoid modification issues if the definition is modified later.
				defCopy := make([]string, len(def))
				copy(defCopy, def)
				wordsToProcess = append(defCopy, wordsToProcess...) // Prepend definition words
			} else {
				return fmt.Errorf("%w: %s", ErrUnknownWord, word)
			}
		}
	}
	return nil
}

// Forth is the main entry point for the Forth evaluator.
// It takes a slice of strings, where each string is a line of Forth code.
func Forth(input []string) ([]int, error) {
	state := newForthState()

	for _, line := range input {
		// Preprocess line: handle definitions spanning multiple lines?
		// The tests seem to imply definitions are per line or handled correctly by sequential processing.
		// Let's assume definitions fit within the `eval` logic for now.
		words := strings.Fields(line)
		if err := state.eval(words); err != nil {
			// If an error occurs during evaluation, return immediately.
			return nil, err
		}
	}

	// After processing all lines, check if we are still in defining mode (missing ';')
	if state.defining {
		return nil, ErrInvalidWord // Indicate an unterminated definition
	}

	// Return the final state of the stack
	return state.stack, nil
}

File: python/forth/forth.py
import math

class StackUnderflowError(Exception):
    """Exception raised when Stack is not full enough to perform the operation."""
    pass


def evaluate(input_data):
    """
    Evaluates a list of strings representing Forth commands using an iterative approach.

    Args:
        input_data: A list of strings, each representing a line of Forth code.

    Returns:
        A list representing the final state of the stack.

    Raises:
        StackUnderflowError: If an operation requires more items than are on the stack.
        ZeroDivisionError: If division by zero is attempted.
        ValueError: If an unknown word is encountered or definition is invalid.
    """
    definitions = {}
    stack = []
    defining = False
    new_word_name = None
    new_word_definition = []

    # --- Built-in Operations ---
    def op_add():
        if len(stack) < 2: raise StackUnderflowError("Insufficient number of items in stack")
        b = stack.pop(); a = stack.pop(); stack.append(a + b)
    def op_sub():
        if len(stack) < 2: raise StackUnderflowError("Insufficient number of items in stack")
        b = stack.pop(); a = stack.pop(); stack.append(a - b)
    def op_mul():
        if len(stack) < 2: raise StackUnderflowError("Insufficient number of items in stack")
        b = stack.pop(); a = stack.pop(); stack.append(a * b)
    def op_div():
        if len(stack) < 2: raise StackUnderflowError("Insufficient number of items in stack")
        b = stack.pop(); a = stack.pop()
        if b == 0: raise ZeroDivisionError("divide by zero")
        stack.append(int(a / b)) # Integer division
    def op_dup():
        if not stack: raise StackUnderflowError("Insufficient number of items in stack")
        stack.append(stack[-1])
    def op_drop():
        if not stack: raise StackUnderflowError("Insufficient number of items in stack")
        stack.pop()
    def op_swap():
        if len(stack) < 2: raise StackUnderflowError("Insufficient number of items in stack")
        stack[-1], stack[-2] = stack[-2], stack[-1]
    def op_over():
        if len(stack) < 2: raise StackUnderflowError("Insufficient number of items in stack")
        stack.append(stack[-2])

    builtins = {
        '+': op_add, '-': op_sub, '*': op_mul, '/': op_div,
        'dup': op_dup, 'drop': op_drop, 'swap': op_swap, 'over': op_over
    }

    # --- Token Processing ---
    token_stream = []
    for line in input_data:
        token_stream.extend(line.split()) # Split all lines into a single token stream

    idx = 0
    while idx < len(token_stream):
        token = token_stream[idx]
        token_lower = token.lower()

        if defining:
            # --- Handling Definition Mode ---
            if token == ';':
                if new_word_name is None:
                    # Should not happen if ':' logic is correct
                    raise ValueError("Invalid definition state: missing name before ';'")

                # Store the definition (lowercase words, numbers as strings)
                processed_definition = []
                for def_token in new_word_definition:
                    try:
                        # Check if it's a number, store as string if so
                        int(def_token)
                        processed_definition.append(def_token)
                    except ValueError:
                        # Store word as lowercase
                        processed_definition.append(def_token.lower())

                definitions[new_word_name] = processed_definition # new_word_name is already lowercase
                defining = False
                new_word_name = None
                new_word_definition = []
                idx += 1 # Consume ';'
            else:
                # Add token to the raw definition list being built
                new_word_definition.append(token)
                idx += 1 # Consume token within definition

        elif token == ':':
            # --- Entering Definition Mode ---
            defining = True
            idx += 1 # Move past ':'
            if idx >= len(token_stream):
                raise ValueError("Invalid definition: missing word name after ':'")

            new_word_name_token = token_stream[idx]
            # Check if trying to define a number (illegal operation)
            try:
                int(new_word_name_token)
                # If int() succeeds, it's an illegal redefinition of a number
                raise ValueError("illegal operation: cannot redefine numbers")
            except ValueError:
                # It's not a number, proceed to define the word
                new_word_name = new_word_name_token.lower() # Store definition name lowercase
                new_word_definition = [] # Start new definition list
                idx += 1 # Consume word name

        elif token_lower in definitions:
            # --- Executing User-Defined Word ---
            # Get a copy of the definition tokens
            definition_tokens = definitions[token_lower][:]
            # Replace the current token (the word name) with its definition
            token_stream[idx:idx+1] = definition_tokens
            # Do not increment idx, the loop will now process the first token
            # of the expanded definition in the next iteration.

        elif token_lower in builtins:
            # --- Executing Built-in Word ---
            builtins[token_lower]()
            idx += 1 # Consume built-in word token

        else:
            # --- Handling Numbers ---
            try:
                stack.append(int(token))
                idx += 1 # Consume number token
            except ValueError:
                # If it's not a definition, not built-in, and not a number, it's undefined
                raise ValueError(f"undefined operation: {token}")


    # After processing all tokens, check if still defining (missing ';')
    if defining:
        raise ValueError("Invalid definition: definition not closed with ';'")

    return stack
